src/de.rs

```diff
@@ -81,7 +102,9 @@ impl<'de, R: Read> Deserializer<R> {
     fn parse_size_information(&mut self, first: u8) -> Result<Option<usize>> {
         let n = self.parse_additional_information(first)?;
         match n {
-            Some(n) if n > MAX_SEQ_LEN => return Err(Error::Syntax),
+            Some(n) if n > self.max_seq_len as u64 => {
```

Comment:
Hm, I wanted to ask why. But I see that in serde that `size_hint` is already capped at 4096, so yes it might not make sense for sequences. And with this in mind, your proposal of growable buffer for strings and bytes is simpler to implement too.